GS-284: Merge users date displays for once only courses

ducthanh-nguyen opened the pull request:
## Description:
- Ensure that the completion record from the old account is transferred to the new account when merging users.
- Update course completion with the latest course completion date for once-only courses, while retaining all previous completions for historical reference.

## Checklist before requesting a review:
<!-- Remove non-applicable items e.g. epic sub-issue points -->
<!-- Add an 'x' inside the brackets to check the item -->

- [x] I have performed a self review of my code.
- [x] My code follows the [Moodle Coding Style](https://moodledev.io/general/development/policies/codingstyle).
- [x] Code has been commented, particularly in hard-to-understand areas.
- [x] Required changes have been made to the documentation (excluding epic sub-issues).
- [x] Any new test cases have been created and existing test cases updated (excluding epic sub-issues).

Commit:
GS-284: - Create class coursecompletionmerger to override generic merger

Commit:
GS-284: - Delete file and replace with class

Commit:
GS-284: Course Completion Table Merger:
- Handle logic for course completion and old completion. I intentionally keep duplicated cases for logic reviewing purpose.
- Add action logs into language file and implementation in config

NickLambellGCH commented:
@ducthanh-nguyen it looks like you're missing a description on this one. There's also a few non-checked tasks.
It's actually good that you didn't bump the version number since we avoid doing that for forked plugins, but you can remove the line so that it isn't counted as not done.

NickLambellGCH commented on lang/en/tool_mergeusers.php:
```diff
@@ -194,3 +194,12 @@
 $string['privacy:metadata'] = 'The Merge User Accounts plugin does not store any personal data.';
+
+$string['mergeusers_completion_updated'] = 'Course completion for course {$a->courseid} updated from user {$a->fromid} to user {$a->toid}.';
+$string['mergeusers_completion_skipped'] = 'Course completion for course {$a->courseid} already exists for user {$a->toid}. Skipping merge.';
```
This string appears to be unused

NickLambellGCH commented on lang/en/tool_mergeusers.php:
```diff
@@ -194,3 +194,12 @@
+$string['mergeusers_recompletion_moved'] = 'Old course completion for user {$a->fromid} and course {$a->courseid} moved to local recompletion.';
+$string['mergeusers_recompletion_error'] = 'Error moving course completion for user {$a->fromid} and course {$a->courseid} to local recompletion: {$a->error}.';
+
+$string['coursecompletionaction'] = 'Course completion action';
+$string['coursecompletionaction_desc'] = 'Description of the course completion action.';
```
These seem to be unused as well

ducthanh-nguyen replied:
I missed pushing the settings file

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+     *
+     * @param array $data array with the necessary data for merging records.
+     * @param array $actionLog list of actions performed.
+     * @param array $errorMessages list of error messages.
```
These can be updated to more specifically define the params as string arrays with `string[]` instead of just `array`.
I try to avoid the generic `array` where possible with the exception being cases like `$data` here where it is actually just an array of arbitrary data rather than a list of items.
I won't comment on all the occurrences, but I'll get you to make the changes to the whole file.

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+<?php
+
```
Looks like this is missing the "This file is part of Moodle ..." license blurb at the top. Not critical but good for consistency 😄

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+     * @param array $actionLog list of actions performed.
+     * @param array $errorMessages list of error messages.
+     */
+    public function merge($data, &$actionLog, &$errorMessages) {
```
Can I get you to add param and return types here and to the other functions in this class?

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+     * Handles the logic for cases where both users have course completions.
+     *
+     * @param object $completion Course completion record for the old user.
+     * @param object $existingCompletion Course completion record for the new user.
```
These should be `stdClass` rather than `object`

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+    protected function handle_existing_completion($completion, $existingCompletion, $fromid, $toid, &$actionLog, &$errorMessages) {
+        global $DB;
+
+        // Both have timestamps: keep the latest, move the older to recompletion.
+        if (!empty($completion->timecompleted) && !empty($existingCompletion->timecompleted)) {
+            if ($completion->timecompleted > $existingCompletion->timecompleted) {
+                // Keep fromid's (latest), move toid's to recompletion.
+                $this->move_to_recompletion($existingCompletion, $toid, $actionLog, $errorMessages);
+                $completion->userid = $toid;
+                $DB->update_record('course_completions', $completion);
+                $actionLog[] = get_string('mergeusers_completion_updated', 'tool_mergeusers', (object)[
+                    'courseid' => $completion->course,
+                    'fromid' => $fromid,
+                    'toid' => $toid
+                ]);
+            }
+
+            if ($completion->timecompleted <= $existingCompletion->timecompleted) {
+                // Keep toid's, move fromid's to recompletion.
+                $this->move_to_recompletion($completion, $fromid, $actionLog, $errorMessages);
+            }
+        }
+
+        // fromid has timestamp, toid has none: move fromid's to recompletion.
+        if (!empty($completion->timecompleted) && empty($existingCompletion->timecompleted)) {
+            $this->move_to_recompletion($completion, $fromid, $actionLog, $errorMessages);
+        }
+
+        // Neither have timestamps: move fromid's to recompletion.
+        if (empty($completion->timecompleted) && empty($existingCompletion->timecompleted)) {
+            $this->move_to_recompletion($completion, $fromid, $actionLog, $errorMessages);
+        }
+
+        // fromid has no timestamp, toid has: move fromid's to recompletion.
+        if (empty($completion->timecompleted) && !empty($existingCompletion->timecompleted)) {
+            $this->move_to_recompletion($completion, $fromid, $actionLog, $errorMessages);
+        }
```
I think this logic might need a bit of a rework. My main indicator for this is that 4 of the possible code paths make exactly the same `move_to_recompletion(...)` call.

I was thinking through it a bit and came to this pseudo code logic flow which might work. Let me know if you have any thoughts on it 😄 
```
if both no timestamp {
	return
}

if existing completion no timestamp {
	delete existing completion?
	transfer new completion
	
	return
}

if new completion < existing completion {
	send new completion to archive
	
	return
}

archive existing completion
transfer new completion
```

ducthanh-nguyen replied:
For existing completion no timestamp, although a completion 'timecompleted' is null, 'timestarted' is different which might be used for reference. Just wondering if we should keep this and an older 'timestarted' might be more valuable (in most cases, existing completion).

NickLambellGCH replied:
I just checked how recompletion handles non-completed courses and there's a bit of a mix. The scheduled task periodic reset skips users with non-completion but manual resets don't check at all and will archive non-completion.
Like you say though, since we're merging users, it's probably good to have the archive as reference. Since that's the case I think it actually simplifies the logic as we can just remove the 'if existing completion no timestamp' part and let it flow through 😄

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+        global $DB;
+
+        // Transfer completion from old user to new user.
+        $completion->userid = $toid;
```
You should always be careful modifying parameters. As a general rule I always avoid it since PHP has somewhat interesting handling sometimes.
In this case, since the param is a stdClass, any modifications to its properties actually flow back into the variable that was used to call the function.

If I want to modify an input param stdClass , I'll always clone it and then use the clone. e.g.
```php
$updatecompletion = clone $completion;
$updatecompletion->userid = $doid;
```
The only catch to note is that this is a shallow clone, so if a property of the object is another object, that object will still be by reference. There are ways to do a deep recursive clone if we ever run into that case but luckily most of our objects are single level 😄

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+        global $DB;
+
+        // Create a record in the local recompletion table (e.g., local_recompletion) for historical purposes.
+        $recompletionData = new stdClass();
```
`$recompletionData` should be `$recompletiondata` with no caps

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+            $actionLog[] = get_string('mergeusers_recompletion_moved', 'tool_mergeusers', (object)[
+                'courseid' => $completion->course,
+                'userid' => $userid
+            ]);
```
New line after the last line in a block where it has content (`catch` in this case) directly after is good

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+        // Delete old user's course completion records.
+        $DB->delete_records('course_completions', ['userid' => $fromid]);
+
+        $actionLog[] = get_string('mergeusers_completion_removed', 'tool_mergeusers', (object)[
+            'fromid' => $fromid
+        ]);
```
The length / splitting of this call is a bit awkward as can happen with `get_string`. Usually, I'll just put each param on its own line and since the data array only has a single entry, it can be collapsed into a single line.

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+                $actionLog[] = get_string('mergeusers_completion_updated', 'tool_mergeusers', (object)[
+                    'courseid' => $completion->course,
+                    'fromid' => $fromid,
+                    'toid' => $toid
```
The last entry in multi-line arrays should have a trailing comma.
There's a couple of these in other functions too 😄

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,164 @@
+            ]);
+
+            // Handle based on completion timestamp presence.
+            if ($existingCompletion) {
+                $this->handle_existing_completion($completion, $existingCompletion, $fromid, $toid, $actionLog, $errorMessages);
+            }
+
+            if (!$existingCompletion) {
```
Logic where you have something like `if($x)` immediately followed by `if(!$x)` usually indicates that the logic could be improved a bit.
In the least optimal cases you can at least change it to an `if ... else` but if you're inside a control structure like a function or loop like this case, you can do it even better.
In this case you should just `continue` after your logic in `if ($existingCompletion)`. Then completely remove the `if (!$existingCompletion)` and place the logic directly in the loop since it will never reach that point if `$existingCompletion` is truthy.

NickLambellGCH reviewed (changes requested):
Thanks for the changes @ducthanh-nguyen, I've given my initial review pass comments (sorry it took so long) 😄 

As a general case that I noticed while reading through the logic. It feels like at some point we may need to delete the 'course_completions' record for `$toid` otherwise when `transfer_completion()` moves records we'd run into a unique key integrity constraint.
I haven't tested it yet, but it just seemed like something that might be an issue.

ducthanh-nguyen committed:
GS-284: Update completion handling logic

- Logic rework in merge function in coursecompletiontablemerger.php to handle completion
- Add actionlog, language file and fix code format
- Settings file updated

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+    /**
+     * @var string current defined action.
+     */
+    private mixed $action;
```
It looks like this property is never used.
The `mixed` type isn't a real type either but rather just one that can be used in PHPDoc. In cases like this where you are storing the result of a `mixed` return type function you can either leave off the prop type or where you should have a known return type like with `get_config()`, you can cast the return value.

NickLambellGCH commented on settings.php:
```diff
@@ -90,6 +90,11 @@
         get_string('uniquekeynewidtomaintain_desc', 'tool_mergeusers'),
         1));
 
+    $settings->add(new admin_setting_configcheckbox('tool_mergeusers/coursecompletionaction',
+        get_string('coursecompletionaction', 'tool_mergeusers'),
+        get_string('coursecompletionaction_desc', 'tool_mergeusers'),
+        1));
```
Since the property this setting is read into in `CourseCompletionTableMerger` isn't used, the setting can probably be removed.

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+        $toid = $data['toid'];
+
+        // Fetch all course completions for the old user.
+        $oldCompletions = $DB->get_records('course_completions', ['userid' => $fromid]);
```
Variables should be all lower case where possible. The only ones that can't be here are the ones passed into the function since it's an override.

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+                $actionLog[] = "Transferring completion for course {$completion->course} to user {$toid}.";
+
+                $this->handle_existing_completion($completion, $existingCompletion, $fromid, $toid, $actionLog, $errorMessages);
+                continue;
```
Should be a new line before keywords like `continue`

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+            (object) ['courseid' => $completion->course]
+        );
+
+        if (empty($completion->timecompleted) && empty($existingCompletion->timecompleted)) {
```
`if` statements with multiple conditions should be expanded to be multi-line

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+            return;
+        }
+
+        if (!empty($completion->timecompleted) &&
+            (empty($existingCompletion->timecompleted) || $completion->timecompleted > $existingCompletion->timecompleted)) {
```
Should be made fully multi-line as above

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+
+        if (empty($completion->timecompleted) && empty($existingCompletion->timecompleted)) {
+            $actionLog[] = "Skipped merge for course {$completion->course} as both completions have no timestamp.";
+            return;
```
Should be a new line before `return`

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+    protected function move_to_recompletion($completion, $userid, &$actionLog, &$errorMessages): void {
+        global $DB;
+
+        $actionLog[] = "Moving course completion (Course ID: {$completion->course}) to recompletion for user {$userid}.";
```
Should this be a language string?

NickLambellGCH commented on lang/en/tool_mergeusers.php:
```diff
@@ -194,3 +194,17 @@
 $string['timetaken'] = 'Merge took {$a} seconds';
 $string['privacy:metadata'] = 'The Merge User Accounts plugin does not store any personal data.';
+
+$string['mergeusers_processing_completion'] = 'Processing completion for course ID {$a->courseid}.';
+$string['mergeusers_existing_found'] = 'Transferring completion for course {$a->courseid} to user {$a->toid}.';
+$string['mergeusers_handling_conflict'] = 'Handling existing completion conflict for course {$a->courseid}.';
+$string['mergeusers_both_empty'] = 'Skipped merge for course {$a->courseid} as both completions have no timestamp.';
+$string['mergeusers_existing_to_recompletion'] = 'Existing completion moved to recompletion and newer completion transferred for course {$a->courseid}.';
+$string['mergeusers_old_to_recompletion'] = 'Old user\'s completion moved to recompletion for course {$a->courseid}.';
+$string['mergeusers_completion_updated'] = 'Updated completion for course {$a->courseid} from user {$a->fromid} to user {$a->toid}.';
+$string['mergeusers_moving_to_recompletion'] = 'Moving course completion (Course ID: {$a->courseid}) to recompletion for user {$a->userid}.';
+$string['mergeusers_recompletion_moved'] = 'Moved course {$a->courseid} completion for user {$a->userid} to recompletion.';
+$string['mergeusers_recompletion_error'] = 'Error moving course {$a->courseid} completion for user {$a->userid} to recompletion: {$a->error}.';
```
Language strings typically shouldn't need the name of the plugin, 'mergeusers_' prefix in this case, as it's already unique to the file and referenced by the plugin in any `get_string()` calls.
In most cases, they also shouldn't have underscores. The main exception being '_help' and '_desc'.

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -0,0 +1,216 @@
+<?php
+
```
Minor one but something the inspection I'm working on integrating into the project picked up on. There shouldn't be a new line between the PHP open tag and the license blurb

ducthanh-nguyen committed:
GS-284: Format fix

- Remove code redundancy
- Fix language string & Fix indentation

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -25,12 +24,6 @@
-
     public function __construct()
     {
         $this->action = get_config('tool_mergeusers', 'coursecompletionaction');
```
Since the 'coursecompletionaction' setting no longer exists, the entire constructor should be removed.

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -99,21 +100,29 @@ protected function handle_existing_completion($completion, $existingCompletion,
 
-        if (empty($completion->timecompleted) && empty($existingCompletion->timecompleted)) {
-            $actionLog[] = "Skipped merge for course {$completion->course} as both completions have no timestamp.";
+        if (empty($completion->timecompleted) &&
+            empty($existingCompletion->timecompleted)) {
```
When expanding an `if` statement to be multi-line, each condition should be on its own line separate from the opening and closing parenthesis

NickLambellGCH commented on lib/table/coursecompletiontablemerger.php:
```diff
@@ -99,21 +100,29 @@ protected function handle_existing_completion($completion, $existingCompletion,
         if (!empty($completion->timecompleted) &&
-            (empty($existingCompletion->timecompleted) || $completion->timecompleted > $existingCompletion->timecompleted)) {
+            (empty($existingCompletion->timecompleted) ||
+                $completion->timecompleted > $existingCompletion->timecompleted)) {
```
This `if` will need some reformatting too. The condition group should be expanded out fully as well. e.g.
```php
if (
    !empty($completion->timecompleted) &&
    (
        empty($existingCompletion->timecompleted) ||
        $completion->timecompleted > $existingCompletion->timecompleted
    )
) {
```

NickLambellGCH commented on lang/en/tool_mergeusers.php:
```diff
@@ -195,16 +195,17 @@
+$string['recompletionerror'] = 'Error moving course {$a->courseid} completion for user {$a->userid} to recompletion: {$a->error}.';
 
 $string['coursecompletionaction'] = 'Course completion action';
 $string['coursecompletionaction_desc'] = 'Description of the course completion action.';
```
These strings should be removed since the setting was removed

ducthanh-nguyen committed:
GS-284: Reformat & delete unnecessary language string
